src/main/java/com/hollingsworth/arsnouveau/common/entity/goal/amethyst_golem/HarvestClusterGoal.java

```diff
@@ -49,28 +60,107 @@ public void tick() {
+            harvestableList.remove(harvested);
+        }
     }
 
-    public void harvest(BlockPos p) {
-        if (!(golem.level instanceof ServerLevel level)) return;
-        for (Direction d : Direction.values()) {
-            BlockState state = level.getBlockState(p.relative(d));
+    // The tool we use to simulate breaking the clusters with.
+    public static final ItemStack TOOL = new ItemStack(Items.DIAMOND_PICKAXE);
+    static {
+        TOOL.set(DataComponents.UNBREAKABLE, new Unbreakable(false));
+    }
+
+    // Reusable list to store drops so that we don't have to keep allocating one for every face or budding block.
+    private final List<ItemStack> drops = new ArrayList<>();
+
+    public boolean harvest(BlockPos p) {
+        if (!(golem.level instanceof ServerLevel level)) return false;
+
+        LootParams.Builder lootBuilder = new LootParams.Builder(level)
+                .withParameter(LootContextParams.TOOL, TOOL);
+
+        boolean shouldDropAsItems = golem.getHome() == null || !golem.canBreak(golem.getHome());
+
+        boolean harvestedAny = false;
+        nextFace: for (Direction d : Direction.values()) {
+            BlockPos pos = p.relative(d);
+            if (!golem.canBreak(pos)) {
+                continue;
+            }
+
+            BlockState state = level.getBlockState(pos);
             if (state.is(CLUSTER_BLOCKS)) {
-                ItemStack stack = new ItemStack(Items.DIAMOND_PICKAXE);
-                state.getBlock().playerDestroy(level, ANFakePlayer.getPlayer(level), p.relative(d), state, level.getBlockEntity(p), stack);
-                destroyBlockSafely(level, p.relative(d), false, ANFakePlayer.getPlayer(level));
+                Vec3 center = Vec3.atCenterOf(pos);
+                Block block = state.getBlock();
+
+                // Clear our reusable list to store new drops.
+                drops.clear();
+
+                // Get the loot table of the block
+                ResourceKey<LootTable> tableKey = block.getLootTable();
+                if (tableKey != BuiltInLootTables.EMPTY) {
+                    LootParams params = lootBuilder
+                            .withParameter(LootContextParams.ORIGIN, center)
+                            .withParameter(LootContextParams.BLOCK_STATE, state)
+                            .create(LootContextParamSets.BLOCK);
+
+                    LootTable table = level.getServer().reloadableRegistries().getLootTable(tableKey);
+                    drops.addAll(table.getRandomItems(params));
+                }
+
+                // Handle drops
+                if (shouldDropAsItems) {
+                    for (ItemStack drop : drops) {
+                        level.addFreshEntity(new ItemEntity(level, pos.getX(), pos.getY(), pos.getZ(), drop));
+                    }
+                } else {
+                    IItemHandler iItemHandler = golem.level.getCapability(Capabilities.ItemHandler.BLOCK, golem.getHome(), null);
```

Comment:
This itemstack logic to insert or drop is handled by InventoryManager

Reply:
slightly different semantics here, we want to insert the full stack or do nothing at all, unless the golem has no home or isnt allowed to access its home